File: pipeline/locate.py
"""Stage 3 - locate the two top-outer corner ROIs the classifier scores.

Two localization paths (whichever produces scorable corners wins):

  * FACE path (worn glasses): a Haar face box (pipeline/facedet) anchors a small
    grid of candidate corner crops per side; features scores them all and keeps
    the peak P(camera). Pure CV can't find glasses on a face, so the face is the
    anchor. Placement imprecision is absorbed by searching the grid.
  * HELD path (glasses held up / studio): the original lens-hole method - split
    the frame bbox at its centre, take the largest interior hole per half as the
    lens, and anchor one corner ROI outside each lens on the end piece.

`two_lens_gate` provides the held-path domain gate (a plausible two-lens frame);
the face path is gated simply by "a face was found".
"""
from __future__ import annotations

from dataclasses import dataclass, field

import cv2
import numpy as np

from config import Config
from pipeline.segment import Segment
from pipeline.facedet import Face


@dataclass
class Lens:
    cx: float
    cy: float
    hw: float   # half-width
    hh: float   # half-height


@dataclass
class CornerROI:
    side: str            # "L" or "R"
    x: int               # ROI top-left in canonical pixels
    y: int
    w: int
    h: int
    cam_cx: float        # expected camera centre (canonical px)
    cam_cy: float


@dataclass
class Located:
    path: str                        # "face" | "held"
    candidates: dict                 # {"L": [CornerROI...], "R": [CornerROI...]}
    method: str
    face: Face | None = None
    lens_left: Lens | None = None
    lens_right: Lens | None = None
    r_lens: float = 0.0
    rois: list = field(default_factory=list)   # best/first ROI per side (viz)


# --------------------------------------------------------------------------- #
# FACE path (worn glasses)
# --------------------------------------------------------------------------- #
def _clip_roi(x, y, w, h, side, shape) -> CornerROI:
    H, W = shape[:2]
    x0 = max(0, int(round(x))); y0 = max(0, int(round(y)))
    x1 = min(W, int(round(x + w))); y1 = min(H, int(round(y + h)))
    return CornerROI(side=side, x=x0, y=y0, w=max(0, x1 - x0), h=max(0, y1 - y0),
                     cam_cx=(x0 + x1) / 2.0, cam_cy=(y0 + y1) / 2.0)


def _face_candidates(face: Face, side: str, cfg: Config, shape) -> list:
    """Grid of candidate corner crops in one top-outer region of the face box."""
    out = []
    for sz in cfg.face_search_sizes:
        w = face.w * sz
        h = w
        for dy in cfg.face_search_dy:
            cy = face.y + face.h * dy
            for dx in cfg.face_search_dx:
                if side == "L":
                    x = face.x + face.w * dx
                else:
                    x = face.x + face.w - w - face.w * dx
                out.append(_clip_roi(x, cy - h / 2.0, w, h, side, shape))
    return out


def locate_face(face: Face, cfg: Config, shape) -> Located:
    cand = {"L": _face_candidates(face, "L", cfg, shape),
            "R": _face_candidates(face, "R", cfg, shape)}
    rois = [cand["L"][0], cand["R"][0]]        # placeholder; features picks best
    return Located(path="face", candidates=cand, method=f"haar:{face.method}",
                   face=face, rois=rois)


# --------------------------------------------------------------------------- #
# HELD path (glasses held up / studio) - original lens-hole method
# --------------------------------------------------------------------------- #
def _holes_in_region(gray_eq, seg: Segment, cfg: Config):
    """Return (cx, cy, w, h, area) for interior holes of the frame contour."""
    x, y, w, h = seg.bbox
    crop = gray_eq[y:y + h, x:x + w]
    _, th = cv2.threshold(crop, 0, 255,
                          cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)
    th = cv2.morphologyEx(th, cv2.MORPH_CLOSE, np.ones((5, 5), np.uint8))
    cnts, hier = cv2.findContours(th, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_SIMPLE)
    holes = []
    if hier is None:
        return holes
    hier = hier[0]
    for i, c in enumerate(cnts):
        if hier[i][3] == -1:          # has no parent -> outer contour, skip
            continue
        area = cv2.contourArea(c)
        if area < 0.02 * w * h:        # ignore tiny gaps
            continue
        bx, by, bw, bh = cv2.boundingRect(c)
        holes.append((x + bx + bw / 2.0, y + by + bh / 2.0, bw, bh, area))
    return holes


def _lens_from_hole(hole) -> Lens:
    cx, cy, w, h, _ = hole
    return Lens(cx=cx, cy=cy, hw=w / 2.0, hh=h / 2.0)


def _fallback_lens(seg: Segment, side: str) -> Lens:
    """Rimless / filled fallback: half-mask centroid + fixed extents."""
    x, y, w, h = seg.bbox
    half = seg.frame_mask[y:y + h,
                          x:x + w // 2] if side == "L" else \
        seg.frame_mask[y:y + h, x + w // 2:x + w]
    m = cv2.moments(half)
    if m["m00"] > 0:
        cx = m["m10"] / m["m00"]
        cy = m["m01"] / m["m00"]
    else:
        cx, cy = (w * 0.25, h * 0.5)
    base_x = x if side == "L" else x + w // 2
    return Lens(cx=base_x + cx, cy=y + cy, hw=0.22 * w, hh=0.50 * h)


def _make_roi(lens: Lens, bbox, side: str, cfg: Config, shape) -> CornerROI:
    """Search region OUTSIDE the lens, on the end piece, where the camera lives.

    Runs from the lens's outer edge (small overlap back into the rim) outward,
    but capped at `roi_outer_max * lens.hw` beyond the lens edge so it can't
    slide onto a hand/hair when the frame bbox is large.
    """
    H, W = shape[:2]
    bx, by, bw, bh = bbox
    overlap = cfg.roi_lens_overlap * lens.hw
    outer_cap = cfg.roi_outer_max * lens.hw
    if side == "L":                      # outer edge is the LEFT side
        x0 = max(bx, lens.cx - lens.hw - outer_cap)
        x1 = lens.cx - lens.hw + overlap
    else:                                # outer edge is the RIGHT side
        x0 = lens.cx + lens.hw - overlap
        x1 = min(bx + bw, lens.cx + lens.hw + outer_cap)
    y0 = by
    y1 = lens.cy + cfg.roi_y_below * lens.hh

    # safety: guarantee a minimum width even if lens detection over-reaches
    min_w = cfg.roi_min_w_frac * bw
    if x1 - x0 < min_w:
        if side == "L":
            x1 = x0 + min_w
        else:
            x0 = x1 - min_w

    x0 = max(0, int(round(x0))); y0 = max(0, int(round(y0)))
    x1 = min(W, int(round(x1))); y1 = min(H, int(round(y1)))
    cam_cx = (x0 + x1) / 2.0             # region centre (weak prior only)
    cam_cy = (y0 + y1) / 2.0
    return CornerROI(side=side, x=x0, y=y0, w=x1 - x0, h=y1 - y0,
                     cam_cx=cam_cx, cam_cy=cam_cy)


def locate_held(gray_eq: np.ndarray, seg: Segment, cfg: Config) -> Located:
    x, y, w, h = seg.bbox
    cx_split = x + w / 2.0
    method = "holes"

    holes = _holes_in_region(gray_eq, seg, cfg)
    left_holes = [hl for hl in holes if hl[0] < cx_split]
    right_holes = [hl for hl in holes if hl[0] >= cx_split]

    if left_holes:
        lens_l = _lens_from_hole(max(left_holes, key=lambda z: z[4]))
    else:
        lens_l = _fallback_lens(seg, "L"); method = "holes+fallback"
    if right_holes:
        lens_r = _lens_from_hole(max(right_holes, key=lambda z: z[4]))
    else:
        lens_r = _fallback_lens(seg, "R"); method = "holes+fallback"

    r_lens = float(np.mean([lens_l.hw, lens_r.hw]))
    roi_l = _make_roi(lens_l, seg.bbox, "L", cfg, gray_eq.shape)
    roi_r = _make_roi(lens_r, seg.bbox, "R", cfg, gray_eq.shape)
    return Located(path="held", candidates={"L": [roi_l], "R": [roi_r]},
                   method=method, lens_left=lens_l, lens_right=lens_r,
                   r_lens=r_lens, rois=[roi_l, roi_r])


def locate(gray_eq: np.ndarray, seg: Segment, cfg: Config,
           face: Face | None = None) -> Located:
    """Face box present -> worn path; else held-glasses lens path."""
    if face is not None:
        return locate_face(face, cfg, gray_eq.shape)
    return locate_held(gray_eq, seg, cfg)


# --------------------------------------------------------------------------- #
# Held-path domain gate: a plausible, tilt-tolerant two-lens frame
# --------------------------------------------------------------------------- #
def two_lens_gate(loc: Located, cfg: Config):
    """Return (passed: bool, reason: str) - the domain gate.

    We deliberately do NOT gate on lens geometry: the held-path lens boxes are too
    noisy (real Ray-Bans whose BOTH corners fire 1.00 were being vetoed for
    "implausible separation"), and gating on studio assumptions is exactly what
    made the original detector reject real photos. FP control rests on the
    both-corners classifier rule + the (retrained) discriminative classifier, which
    already holds normal_frame at 25/25 NORMAL. The gate only confirms structure is
    present: a face was anchored, or two lens regions were located.
    """
    if loc.path == "face":
        return (loc.face is not None), ("face anchor" if loc.face else "no face")
    if loc.lens_left is None or loc.lens_right is None:
        return False, "no two lenses found"
    return True, "held glasses frame"

File: static/detector/locate.js
/* Stage 3 - locate the two top-outer corner ROIs the classifier scores.
 * Port of pipeline/locate.py. Two paths (whichever yields scorable corners wins):
 *
 *   FACE path (worn glasses): a Haar face box anchors a small grid of candidate
 *     corner crops per side; features scores them all and keeps the peak P(camera).
 *   HELD path (glasses held / studio): the lens-hole method - split the frame bbox
 *     at its centre, take the largest interior hole per half as the lens, and anchor
 *     one corner ROI outside each lens on the end piece.
 *
 * locate() returns candidates:{L:[...],R:[...]}; the caller peaks per side.
 * twoLensGate() is the domain gate: a face was anchored, or two lenses were found.
 */
(function (root, factory) {
  var mod = factory();
  if (typeof module !== 'undefined' && module.exports) { module.exports = mod; }
  if (root) { root.DET = root.DET || {}; root.DET.locate = mod; }
})(typeof self !== 'undefined' ? self : (typeof window !== 'undefined' ? window : null),
function () {
  'use strict';

  // ------------------------------------------------------------------ //
  // FACE path (worn glasses)
  // ------------------------------------------------------------------ //
  function clipRoi(x, y, w, h, side, H, W) {
    var x0 = Math.max(0, Math.round(x)), y0 = Math.max(0, Math.round(y));
    var x1 = Math.min(W, Math.round(x + w)), y1 = Math.min(H, Math.round(y + h));
    return {
      side: side, x: x0, y: y0, w: Math.max(0, x1 - x0), h: Math.max(0, y1 - y0),
      camCx: (x0 + x1) / 2, camCy: (y0 + y1) / 2
    };
  }

  /** Grid of candidate corner crops in one top-outer region of the face box. */
  function faceCandidates(face, side, cfg, H, W) {
    var out = [];
    var si, di, dj;
    for (si = 0; si < cfg.face_search_sizes.length; si += 1) {
      var w = face.w * cfg.face_search_sizes[si];
      var h = w;
      for (di = 0; di < cfg.face_search_dy.length; di += 1) {
        var cy = face.y + face.h * cfg.face_search_dy[di];
        for (dj = 0; dj < cfg.face_search_dx.length; dj += 1) {
          var dx = cfg.face_search_dx[dj];
          var x = side === 'L'
            ? face.x + face.w * dx
            : face.x + face.w - w - face.w * dx;
          out.push(clipRoi(x, cy - h / 2, w, h, side, H, W));
        }
      }
    }
    return out;
  }

  function locateFace(face, cfg, H, W) {
    var cand = { L: faceCandidates(face, 'L', cfg, H, W),
                 R: faceCandidates(face, 'R', cfg, H, W) };
    return {
      path: 'face', candidates: cand, method: 'haar:' + face.method, face: face,
      lensLeft: null, lensRight: null, rLens: 0,
      rois: [cand.L[0], cand.R[0]]   // placeholder; caller replaces with peak
    };
  }

  // ------------------------------------------------------------------ //
  // HELD path (glasses held up / studio) - lens-hole method
  // ------------------------------------------------------------------ //
  /** Interior holes (lens openings) of the frame contour, in full-image px. */
  function holesInRegion(grayEq, seg, cv) {
    var x = seg.bbox[0], y = seg.bbox[1], w = seg.bbox[2], h = seg.bbox[3];
    var crop = grayEq.roi(new cv.Rect(x, y, w, h));
    var th = new cv.Mat();
    cv.threshold(crop, th, 0, 255, cv.THRESH_BINARY_INV + cv.THRESH_OTSU);
    var k5 = cv.getStructuringElement(cv.MORPH_RECT, new cv.Size(5, 5));
    cv.morphologyEx(th, th, cv.MORPH_CLOSE, k5, new cv.Point(-1, -1), 1,
      cv.BORDER_CONSTANT, cv.morphologyDefaultBorderValue());
    k5.delete();

    var contours = new cv.MatVector();
    var hier = new cv.Mat();
    cv.findContours(th, contours, hier, cv.RETR_CCOMP, cv.CHAIN_APPROX_SIMPLE);
    var holes = [];
    var par = hier.data32S; // [next,prev,child,parent] per contour
    var i;
    for (i = 0; i < contours.size(); i += 1) {
      if (par[i * 4 + 3] === -1) { continue; } // outer contour -> skip
      var area = cv.contourArea(contours.get(i));
      if (area < 0.02 * w * h) { continue; }
      var b = cv.boundingRect(contours.get(i));
      holes.push([x + b.x + b.width / 2, y + b.y + b.height / 2, b.width, b.height, area]);
    }
    contours.delete(); hier.delete(); th.delete(); crop.delete();
    return holes;
  }

  function lensFromHole(hole) {
    return { cx: hole[0], cy: hole[1], hw: hole[2] / 2, hh: hole[3] / 2 };
  }

  function fallbackLens(seg, side, cv) {
    var x = seg.bbox[0], y = seg.bbox[1], w = seg.bbox[2], h = seg.bbox[3];
    var halfW = side === 'L' ? Math.floor(w / 2) : w - Math.floor(w / 2);
    var halfX = side === 'L' ? x : x + Math.floor(w / 2);
    var half = seg.frameMask.roi(new cv.Rect(halfX, y, halfW, h));
    var m = cv.moments(half, false);
    half.delete();
    var cx, cy;
    if (m.m00 > 0) { cx = m.m10 / m.m00; cy = m.m01 / m.m00; }
    else { cx = w * 0.25; cy = h * 0.5; }
    var baseX = side === 'L' ? x : x + Math.floor(w / 2);
    return { cx: baseX + cx, cy: y + cy, hw: 0.22 * w, hh: 0.50 * h };
  }

  /* Search region OUTSIDE the lens, on the end piece, where the camera lives.
     Capped at roi_outer_max * lens.hw beyond the lens edge so it can't slide onto
     a hand/hair when the frame bbox is large (port of _make_roi). */
  function makeRoi(lens, bbox, side, cfg, H, W) {
    var bx = bbox[0], by = bbox[1], bw = bbox[2];
    var overlap = cfg.roi_lens_overlap * lens.hw;
    var outerCap = cfg.roi_outer_max * lens.hw;
    var x0, x1;
    if (side === 'L') {
      x0 = Math.max(bx, lens.cx - lens.hw - outerCap);
      x1 = lens.cx - lens.hw + overlap;
    } else {
      x0 = lens.cx + lens.hw - overlap;
      x1 = Math.min(bx + bw, lens.cx + lens.hw + outerCap);
    }
    var y0 = by;
    var y1 = lens.cy + cfg.roi_y_below * lens.hh;

    var minW = cfg.roi_min_w_frac * bw;
    if (x1 - x0 < minW) {
      if (side === 'L') { x1 = x0 + minW; } else { x0 = x1 - minW; }
    }
    x0 = Math.max(0, Math.round(x0)); y0 = Math.max(0, Math.round(y0));
    x1 = Math.min(W, Math.round(x1)); y1 = Math.min(H, Math.round(y1));
    return {
      side: side, x: x0, y: y0, w: x1 - x0, h: y1 - y0,
      camCx: (x0 + x1) / 2, camCy: (y0 + y1) / 2
    };
  }

  function locateHeld(grayEq, seg, cfg, cv) {
    var H = grayEq.rows, W = grayEq.cols;
    var x = seg.bbox[0], w = seg.bbox[2];
    var cxSplit = x + w / 2;
    var method = 'holes';

    var holes = holesInRegion(grayEq, seg, cv);
    var left = holes.filter(function (z) { return z[0] < cxSplit; });
    var right = holes.filter(function (z) { return z[0] >= cxSplit; });

    function maxByArea(arr) {
      return arr.reduce(function (a, b) { return b[4] > a[4] ? b : a; });
    }

    var lensL, lensR;
    if (left.length) { lensL = lensFromHole(maxByArea(left)); }
    else { lensL = fallbackLens(seg, 'L', cv); method = 'holes+fallback'; }
    if (right.length) { lensR = lensFromHole(maxByArea(right)); }
    else { lensR = fallbackLens(seg, 'R', cv); method = 'holes+fallback'; }

    var rLens = (lensL.hw + lensR.hw) / 2;
    var roiL = makeRoi(lensL, seg.bbox, 'L', cfg, H, W);
    var roiR = makeRoi(lensR, seg.bbox, 'R', cfg, H, W);
    return {
      path: 'held', candidates: { L: [roiL], R: [roiR] }, method: method,
      face: null, lensLeft: lensL, lensRight: lensR, rLens: rLens,
      rois: [roiL, roiR]
    };
  }

  /**
   * Face box present -> worn path; else held-glasses lens path.
   * @param {cv.Mat} grayEq CV_8UC1
   * @param {object} seg  segment() result
   * @param {object} cfg
   * @param {object} cv
   * @param {object|null} [face]  facedet.detectFace() result
   */
  function locate(grayEq, seg, cfg, cv, face) {
    if (face) { return locateFace(face, cfg, grayEq.rows, grayEq.cols); }
    return locateHeld(grayEq, seg, cfg, cv);
  }

  /**
   * Domain gate: confirm structure is present. Face anchored, or two lenses found.
   * @returns {{ok:boolean, reason:string}}
   */
  function twoLensGate(loc) {
    if (loc.path === 'face') {
      return { ok: !!loc.face, reason: loc.face ? 'face anchor' : 'no face' };
    }
    if (!loc.lensLeft || !loc.lensRight) {
      return { ok: false, reason: 'no two lenses found' };
    }
    return { ok: true, reason: 'held glasses frame' };
  }

  return { locate: locate, twoLensGate: twoLensGate };
});
